moveit_planners/trajopt/src/kinematic_terms.cpp

```diff
@@ -0,0 +1,67 @@
+#include <Eigen/Geometry>
+#include <boost/format.hpp>
+
+#include <trajopt_sco/expr_ops.hpp>
+#include <trajopt_sco/modeling_utils.hpp>
+
+#include "trajopt_interface/kinematic_terms.h"
+
+using namespace std;
+using namespace sco;
+using namespace Eigen;
+
+namespace trajopt_interface
+{
+VectorXd CartPoseErrCalculator::operator()(const VectorXd& dof_vals) const
+{
+  // TODO: create the actual error function from information in planning scene
+  VectorXd err;
+  return err;
+}
+
+VectorXd JointVelErrCalculator::operator()(const VectorXd& var_vals) const
+{
+  assert(var_vals.rows() % 2 == 0);
+  // Top half of the vector are the joint values. The bottom half are the 1/dt values
+  int half = static_cast<int>(var_vals.rows() / 2);
+  int num_vels = half - 1;
+  // (x1-x0)*(1/dt)
+  VectorXd vel = (var_vals.segment(1, num_vels) - var_vals.segment(0, num_vels)).array() *
+                 var_vals.segment(half + 1, num_vels).array();
+
+  // Note that for equality terms tols are 0, so error is effectively doubled
+  VectorXd result(vel.rows() * 2);
+  result.topRows(vel.rows()) = -(upper_tol_ - (vel.array() - target_));
+  result.bottomRows(vel.rows()) = lower_tol_ - (vel.array() - target_);
+  return result;
+}
+
+MatrixXd JointVelJacobianCalculator::operator()(const VectorXd& var_vals) const
+{
+  // var_vals = (theta_t1, theta_t2, theta_t3 ... 1/dt_1, 1/dt_2, 1/dt_3 ...)
```

Comment:
These two comments are describing the same thing in very different ways. IMO I like the latter better, can we stick to that?

Reply:
sorry, I did not get it. which two comments?